Better typing for overloaded higher-order methods

adriaanm opened the pull request:
TODO: 
  - [x] spec update
  - [ ] performance
  - [x] clean up OverloadedArgFunProto

For better typing for an overloaded higher-order method such as `map` (prevalent in the new collections
/ spark apis), regardless of whether the actual argument is a function literal, a method reference that
will be eta-expanded (based on the expected type), or a method value (`m _`).

Before, we triggered this based on the syntactic shape of the arguments, but that results (in our case)
in difference between `foo`, `foo _` and `x => foo(x)` as arguments to overloaded method.

The first commit (ProtoType generalizes [Bounded]WildcardType) sneakily sets the scene, pretty
specifically for the second one, but I do expect it to be useful for other cases.

Normally, overload resolution types the arguments to the alternatives without an expected type.
However, typing function literals and eta-expansion are driven by the expected type:

  - function literals usually don't have parameter types, which are derived from the expected type;

  - eta-expansion right now only happens when a function/sam type is expected.

(Dotty side-steps these issues by eta-expanding regardless of expected type.)

Now that the collections are full of overloaded HO methods, we should try harder to type check them
nicely.

To avoid breaking existing code, we only provide an expected type (for each argument position) when:

 - there is at least one FunctionN type expected by one of the overloads: in this case, the expected
   type is a FunctionN[Ti, ?], where Ti are the argument types (they must all be =:=), and the expected
   result type is elided using a wildcard. This does not exclude any overloads that expect a SAM,
   because they conform to a function type through SAM conversion

 - OR: all overloads expect a SAM type of the same class, but with potentially varying result types
   (argument types must be =:=)

We allow polymorphic cases, as long as the types parameters are instantiated by the AntiPolyType prefix.

In all other cases, the old behavior is maintained: Wildcard is expected.

(Slightly) more formally:

Consider an overloaded method `m_i`, with `N` overloads `i = 1..N`, and an expected argument type at
index `j`, `a_ij`:

```
def m_1(... a_1j, ...)
..
def m_N(... a_Nj, ...)
```

Any polymorphic method `m_i` will be reduced to the monomorphic case by pushing down the method's
`PolyType` to its arguments `a_ij`.

The expected type for the argument at index `j` will be more precise than the usual `WildcardType`
(`?`), if all types `a_1j..a_Nj` are function-ish types that denote the same parameter types `p1..pM`.

A "function-ish" type is a `FunctionN[p1,...,pM]` (or `PartialFunction`), or the equivalent SAM type.
(We first unwrap any PolyTypes.)

The non-wildcard expected type will be
  - `PartialFunction[p1, ?]`, if an `a_ij` expects a partial function;
  - else, if there is a subclass of `FunctionM` among the `a_ij`, it is `FunctionM[p1, pM, ?]`;
  - else, if all `a_ij` are of the same SAM type (allowing for varying result types), 
    that SAM type (with `?` result type).

In each case, any type parameter not already resolved by overloading, or the outer context, it
approximated by `?`.

PS: type equivalence is decided as `tp1 <:< tp2 && tp2 <:< tp1`, and not `tp1 =:= tp2` (the latter is
actually stricter).

adriaanm commented:
[test] [error] /home/jenkins/workspace/scala-2.13.x-validate-main@2/src/compiler/scala/reflect/quasiquotes/Reifiers.scala:94: overloaded method value collect with alternatives:
[test] [error]   [K2, V2](pf: PartialFunction[(Reifiers.this.global.Name, Set[Reifiers.this.global.TermName]),(K2, V2)])scala.collection.mutable.Map[K2,V2] <and>
[test] [error]   [B](pf: PartialFunction[(Reifiers.this.global.Name, Set[Reifiers.this.global.TermName]),B])scala.collection.mutable.Iterable[B]
[test] [error]  cannot be applied to (((Reifiers.this.global.Name, Set[Reifiers.this.global.TermName])) => Iterator[Reifiers.this.global.Apply])
[test] [error]               nameMap.collect { case (_, nameset) if nameset.size >= 2 =>
[test] [error]                       ^
[test] [error] /home/jenkins/workspace/scala-2.13.x-validate-main@2/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:971: value ++= is not a member of Set[SpecializeTypes.this.global.Symbol]
[test] [error]   Expression does not convert to assignment because:
[test] [error]     overloaded method value collect with alternatives:
[test] [error]       [K2, V2](pf: PartialFunction[(SpecializeTypes.this.global.Symbol, SpecializeTypes.this.global.Type),(K2, V2)])scala.collection.immutable.Map[K2,V2] <and>
[test] [error]       [B](pf: PartialFunction[(SpecializeTypes.this.global.Symbol, SpecializeTypes.this.global.Type),B])scala.collection.immutable.Iterable[B]
[test] [error]      cannot be applied to (((SpecializeTypes.this.global.Symbol, SpecializeTypes.this.global.Type)) => SpecializeTypes.this.global.Symbol)
[test] [error]     expansion: SpecializeTypes.this.wasSpecializedForTypeVars.update(specMember, SpecializeTypes.this.wasSpecializedForTypeVars.apply(specMember).++(spec.<collect: error>(<empty> match {
[test] [error]   case scala.Tuple2((s @ _), (tp @ _)) if s.tpe.$eq$eq(tp) => s
[test] [error] })))
[test] [error]       wasSpecializedForTypeVars(specMember) ++= spec collect { case (s, tp) if s.tpe == tp => s }
[test] [error]                                             ^
[test] [error] /home/jenkins/workspace/scala-2.13.x-validate-main@2/src/compiler/scala/tools/nsc/typechecker/Infer.scala:1433: overloaded method value collect with alternatives:
[test] [error]   [B](pf: PartialFunction[(Infer.this.global.Symbol, Option[Infer.this.global.Type]),B])scala.collection.mutable.Iterable[B] <and>
[test] [error]   [K2, V2](pf: PartialFunction[(Infer.this.global.Symbol, Option[Infer.this.global.Type]),(K2, V2)])scala.collection.mutable.LinkedHashMap[K2,V2]
[test] [error]  cannot be applied to (((Infer.this.global.Symbol, Option[Infer.this.global.Type])) => (Infer.this.global.Symbol, Infer.this.global.Type))
[test] [error]       (m collect {case (p, Some(a)) => (p, a)}).unzip  ))
[test] [error]          ^
[test] [error] three errors found
[test] [error] /home/jenkins/workspace/scala-2.13.x-validate-main@2/src/compiler/scala/reflect/quasiquotes/Reifiers.scala:94: overloaded method value collect with alternatives:
[test] [error]   [K2, V2](pf: PartialFunction[(Reifiers.this.global.Name, Set[Reifiers.this.global.TermName]),(K2, V2)])scala.collection.mutable.Map[K2,V2] <and>
[test] [error]   [B](pf: PartialFunction[(Reifiers.this.global.Name, Set[Reifiers.this.global.TermName]),B])scala.collection.mutable.Iterable[B]
[test] [error]  cannot be applied to (((Reifiers.this.global.Name, Set[Reifiers.this.global.TermName])) => Iterator[Reifiers.this.global.Apply])
[test] [error]               nameMap.collect { case (_, nameset) if nameset.size >= 2 =>
[test] [error]                       ^
[test] [error] /home/jenkins/workspace/scala-2.13.x-validate-main@2/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:971: value ++= is not a member of Set[SpecializeTypes.this.global.Symbol]
[test] [error]   Expression does not convert to assignment because:
[test] [error]     overloaded method value collect with alternatives:
[test] [error]       [K2, V2](pf: PartialFunction[(SpecializeTypes.this.global.Symbol, SpecializeTypes.this.global.Type),(K2, V2)])scala.collection.immutable.Map[K2,V2] <and>
[test] [error]       [B](pf: PartialFunction[(SpecializeTypes.this.global.Symbol, SpecializeTypes.this.global.Type),B])scala.collection.immutable.Iterable[B]
[test] [error]      cannot be applied to (((SpecializeTypes.this.global.Symbol, SpecializeTypes.this.global.Type)) => SpecializeTypes.this.global.Symbol)
[test] [error]     expansion: SpecializeTypes.this.wasSpecializedForTypeVars.update(specMember, SpecializeTypes.this.wasSpecializedForTypeVars.apply(specMember).++(spec.<collect: error>(<empty> match {
[test] [error]   case scala.Tuple2((s @ _), (tp @ _)) if s.tpe.$eq$eq(tp) => s
[test] [error] })))
[test] [error]       wasSpecializedForTypeVars(specMember) ++= spec collect { case (s, tp) if s.tpe == tp => s }
[test] [error]                                             ^
[test] [error] /home/jenkins/workspace/scala-2.13.x-validate-main@2/src/compiler/scala/tools/nsc/typechecker/Infer.scala:1433: overloaded method value collect with alternatives:
[test] [error]   [B](pf: PartialFunction[(Infer.this.global.Symbol, Option[Infer.this.global.Type]),B])scala.collection.mutable.Iterable[B] <and>
[test] [error]   [K2, V2](pf: PartialFunction[(Infer.this.global.Symbol, Option[Infer.this.global.Type]),(K2, V2)])scala.collection.mutable.LinkedHashMap[K2,V2]
[test] [error]  cannot be applied to (((Infer.this.global.Symbol, Option[Infer.this.global.Type])) => (Infer.this.global.Symbol, Infer.this.global.Type))
[test] [error]       (m collect {case (p, Some(a)) => (p, a)}).unzip  ))
[test] [error]          ^
[test] [error] three errors found

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2973,7 +3091,7 @@ trait Types
       val pre1 = if (pre.isInstanceOf[ClassInfoType]) pre else map(pre)
       if (pre1 eq pre) this
-      else OverloadedType(pre1, alternatives)
+      else OverloadedType(pre1, alternatives) // TODO: shouldn't we also map over the alternatives?
```
That's tricky, because we can't just go an clone `alternatives` here, as these symbols are serving double duty of carrying an info and linking to an actual member.

For the particular case of `AsSeenFrom`, we take care to [later call `memberType`](https://github.com/scala/scala/blob/37b3892fa06cf30e0335397a2d020ea757309c37/src/compiler/scala/tools/nsc/typechecker/Infer.scala#L1421) on the eventually selected alternative.

adriaanm committed:
ProtoType generalizes [Bounded]WildcardType

This will serve as a hook into type checking to represent
expectation of types that cannot (cheaply) be encoded as
existing types.

adriaanm committed:
More precise prototype for args of overloaded method

Normally, overload resolution types the arguments to the alternatives
without an expected type. However, typing function literals and
eta-expansion are driven by the expected type:

  - function literals usually don't have parameter types, which are
    derived from the expected type;

  - eta-expansion right now only happens when a function/sam type is
    expected.

(Dotty side-steps these issues by eta-expanding regardless of
expected type.)

Now that the collections are full of overloaded HO methods, we should
try harder to type check them nicely.

To avoid breaking existing code, we only provide an expected type (for
each argument position) when:

 - there is at least one FunctionN type expected by one of the
   overloads: in this case, the expected type is a FunctionN[Ti, ?],
   where Ti are the argument types (they must all be =:=), and the
   expected result type is elided using a wildcard. This does not
   exclude any overloads that expect a SAM, because they conform to a
   function type through SAM conversion

 - OR: all overloads expect a SAM type of the same class, but with
   potentially varying result types (argument types must be =:=)

We allow polymorphic cases, as long as the types parameters are
instantiated by the AntiPolyType prefix.

In all other cases, the old behavior is maintained: Wildcard is
expected.

(Slightly) more formally:

Consider an overloaded method `m_i`, with `N` overloads `i = 1..N`,
and an expected argument type at index `j`, `a_ij`:

```
def m_1(... a_1j, ...)
..
def m_N(... a_Nj, ...)
```

Any polymorphic method `m_i` will be reduced to the monomorphic case
by pushing down the method's `PolyType` to its arguments `a_ij`.

The expected type for the argument at index `j` will be more
precise than the usual `WildcardType` (`?`), if all types `a_1j..a_Nj`
are function-ish types that denote the same parameter types `p1..pM`.

A "function-ish" type is a `FunctionN[p1,...,pM]` (or
`PartialFunction`), or the equivalent SAM type. (We first unwrap any
PolyTypes.)

The non-wildcard expected type will be
  - `PartialFunction[p1, ?]`, if an `a_ij` expects a partial function;
  - else, if there is a subclass of `FunctionM` among the `a_ij`,
    it is `FunctionM[p1, pM, ?]`;
  - else, if all `a_ij` are of the same SAM type (allowing
    for varying result types), that SAM type (with `?` result type).

In each case, any type parameter not already resolved by overloading,
or the outer context, it approximated by `?`.

PS: type equivalence is decided as `tp1 <:< tp2 && tp2 <:< tp1`,
and not `tp1 =:= tp2` (the latter is actually stricter).

adriaanm commented:
@SethTisue could you run a community build on this one?

SethTisue commented:
specs2 fails with:

```
[specs2] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.12/project-builds/specs2-41304f739191bc5c34e6c3a1a87816dcae5ea3c3/common/shared/src/main/scala/org/specs2/data/Trees.scala:48:29: type mismatch;
[specs2] [error]  found   : Option[A(in method clean)] => Option[A(in method clean)]
[specs2] [error]  required: ?<: org.specs2.fp.Tree[?] => ?
[specs2] [error]     prune(t, (a: Option[A]) => a).getOrElse(Leaf(initial))
[specs2] [error]                             ^
[specs2] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.12/project-builds/specs2-41304f739191bc5c34e6c3a1a87816dcae5ea3c3/common/shared/src/main/scala/org/specs2/data/Trees.scala:48:29: type mismatch;
[specs2] [error]  found   : Option[A(in method clean)] => Option[A(in method clean)]
[specs2] [error]  required: ?<: org.specs2.fp.Tree[?] => ?
[specs2] [error]     prune(t, (a: Option[A]) => a).getOrElse(Leaf(initial))
[specs2] [error]                             ^
[specs2] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.12/project-builds/specs2-41304f739191bc5c34e6c3a1a87816dcae5ea3c3/common/shared/src/main/scala/org/specs2/data/Trees.scala:54:49: type mismatch;
[specs2] [error]  found   : A(in method prune) => Option[(some other)B(in method prune)]
[specs2] [error]  required: ?<: org.specs2.fp.Tree[?] => ?
[specs2] [error]     val tbs = t.subForest.flatMap(t => prune(t, f))
[specs2] [error]                                                 ^
[specs2] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.12/project-builds/specs2-41304f739191bc5c34e6c3a1a87816dcae5ea3c3/common/shared/src/main/scala/org/specs2/data/Trees.scala:54:49: type mismatch;
[specs2] [error]  found   : A(in method prune) => Option[(some other)B(in method prune)]
[specs2] [error]  required: ?<: org.specs2.fp.Tree[?] => ?
[specs2] [error]     val tbs = t.subForest.flatMap(t => prune(t, f))
[specs2] [error]                                                 ^
```

niktrop commented:
@retronym Thanks! I'll have a look.

adriaanm commented:
Another shot at community building: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1248/

adriaanm commented:
Next failure is in gigahorse:
```
okhttp/src/main/scala/gigahorse/support/okhttp/OkhClient.scala:51: type mismatch;
 found   : gigahorse.support.okhttp.OkHandler.FullOkHandler[gigahorse.FullResponse]
 required: ?<: gigahorse.FullResponse => ?
    processFull(request, OkHandler[FullResponse](identity))
                                                ^
okhttp/src/main/scala/gigahorse/support/okhttp/OkhClient.scala:55: type mismatch;
 found   : gigahorse.support.okhttp.OkHandler.FullOkHandler[A(in method run)]
 required: ?<: gigahorse.FullResponse => ?
    processFull(request, OkHandler[A](f))
                                     ^
okhttp/src/main/scala/gigahorse/support/okhttp/OkhClient.scala:147: type mismatch;
 found   : gigahorse.support.okhttp.FunctionHandler[gigahorse.FullResponse]
 required: ?<: gigahorse.FullResponse => ?
    processFull(request, FunctionHandler[FullResponse](identity))
                                                      ^
okhttp/src/main/scala/gigahorse/support/okhttp/OkhClient.scala:151: type mismatch;
 found   : gigahorse.support.okhttp.FunctionHandler[A(in method processFull)]
 required: ?<: gigahorse.FullResponse => ?
    processFull(request, FunctionHandler[A](f))
                                           ^
```

@SethTisue could you double check to see if there's another new failure? scodec-bits had a test fail, while it was passing in the last community build that I compared to -- for now assuming that's not related to this PR

SethTisue commented:
I've seen the scodec-bits test failure in other runs, no worries there.

this PR seems to have been made shapeless green when it was red before, was that on purpose?

(shapeless being green adds two failures, case-app and play-json, but I checked and the errors are similar to those from older runs before shapeless regressed, so no worries there)

(also note that as of run 1250 we have moved to a newer SHA, fd820e4. so if you end up wanting to do another community build run, I'd suggest rebasing this PR onto fd820e4 first.)

adriaanm commented:
The shapeless progression could be if it used method values as arguments
for map, as that’s the kind of thing this is trying to fix (before we only
handled syntactically obvious function literals as arguments to overloaded
methods)
On Wed, Jul 18, 2018 at 23:08 Seth Tisue <notifications@github.com> wrote:

> I've seen the scodec-bits test failure in other runs, no worries there.
>
> this PR seems to have been made shapeless green when it was red before,
> was that on purpose?
>
> (shapeless being green adds two failures, case-app and play-json, but I
> checked and the errors are similar to those from older runs before
> shapeless regressed, so no worries there)
>
> (also note that as of run 1250 we have moved to a newer SHA, fd820e4. so
> if you end up wanting to do another community build run, I'd suggest
> rebasing this PR onto fd820e4 first.)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/6871#issuecomment-406075079>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjy3mep6CZ-Ms2kiZKvNAaeuTNg4znks5uH6PigaJpZM4U9BYG>
> .
>

adriaanm commented:
The gigahorse fail boils down to:

```
abstract class Sam[A] { def apply(a: String): A }

class Test {
  def map[A](f: String => A): A = map(new Sam[A] { def apply(a: String): A = f(a) })
  def map[A](f: Sam[A]): A = ???
}
```

🤔 we should probably support that 😄

adriaanm commented:
Note to self (so I remember after vacation :-)): I have some further non-essential WIP to also make TypeVars prototypes at overload_proto_wip.

adriaanm commented:
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1253/

smarter commented:
@adriaanm It'd be nice if we could establish a list of all the test cases related to the overloading resolution changes in 2.13 so that we can try to align dotty with it.

adriaanm commented:
Anything that I’ve worked on should have “overload” in the test file name.
(I’m on vacation until Aug 2)
On Fri, Jul 20, 2018 at 20:24 Guillaume Martres <notifications@github.com>
wrote:

> @adriaanm <https://github.com/adriaanm> It'd be nice if we could
> establish a list of all the test cases related to the overloading
> resolution changes in 2.13 so that we can try to align dotty with it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/6871#issuecomment-406687235>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjy4c9Oyo6PoMZYdPt3uh0pUhuZTPAks5uIiBIgaJpZM4U9BYG>
> .
>

milessabin commented:
shapeless was failing before because of a change in the constructor signature of `ImplicitSearch` in the byname implicits PR ... I think it's highly unlikely that anything happening here could have changed that.

adriaanm commented:
re-review by @retronym?

retronym commented:
I ran this through a [performance test](https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/1644/console) ([results](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&var-scalaSha=53a1121461%7C8f57e5a2f8%7C6ed12b274e%7Cb3646b7d7a%7C8d173dae09%7C224f10d4cb%7Cd71f4da401%7Cae43152e5b%7Ceaeda754e2%7Cfd820e4291%7C9b3a69cecc%7C5292baacac%7Cf099d8813f%7Cb24a6e03a6%7C2de8b62488%7C6ff29f967f%7C4d39d4f849%7Cb6771349e6%7Cdb12e723f7%7C40d0f04830&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40&from=1531082222295&to=1533269654927)).

The `scalap` benchmark looks a touch slower (~1.5%).

The `scala` benchmark failed to compile with:

```
[error] /home/jenkins/.compilerBenchmark/sourceAssembly/reflect/scala/reflect/internal/Types.scala:227: error: overloaded method transformedType needs result type
[error]         var result: Type = transformedType(this)
[error]                                           ^
```

SethTisue commented:
> https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1253/

the results here were the same as run 1254, another 2.13.x run from around the same time, so I'd say we have a clean bill of health.

retronym commented:
I've run the performance tests for the `scala` corpus, too (after editing the sources to avoid what appears to be a reasonable new cyclic error after this PR).

It also sees the ~2% slowdown. Not sure yet how much of that is inherent to the change vs something we can ratchet down (e.g. by caching whether a given class symbol is a functional interface or not)

adriaanm commented:
@retronym are you comfortable with merging this for M5 and working on performance for RC1?

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -629,7 +631,7 @@ trait Implicits {
       case MethodType(p :: _, restpe) if p.isImplicit => matchesPtView(restpe, ptarg, ptres, undet)
       case MethodType(p :: Nil, restpe)               => matchesArgRes(p.tpe, restpe, ptarg, ptres, undet)
-      case ExistentialType(_, qtpe)                   => matchesPtView(normalize(qtpe), ptarg, ptres, undet)
+      case ExistentialType(_, qtpe)                   => matchesPtView(methodToExpressionTp(qtpe), ptarg, ptres, undet)
```
❤️

lrytz commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1192,6 +1192,134 @@ trait Types
+    *   - there is at least one FunctionN type expected by one of the overloads:
+    *     in this case, the expected type is a FunctionN[Ti, ?], where Ti are the argument types (they must all be =:=),
+    *     and the expected result type is elided using a wildcard.
+    *     This does not exclude any overloads that expect a SAM, because they conform to a function type through SAM conversion
```
hmmm.. i though sam conversion works on literals, i.e., a function literal can get a sam type. if the expected type is a Function, isn't the literal typed with a Function type? how can that get converted to a sam type later?

anyway, it works as intended, so ignore my rambling.

```
scala> trait S[-T, +U] { def apply(x: T): U }
defined trait S

scala> object C { def m(f: String S Int) = 0; def m(f: String => String) = 1 }
defined object C

scala> C.m(x => 1)
res5: Int = 0
```

adriaanm replied:
It is first assigned the Function type, but the expected type is still the SAM. That gap is closed during adapt, case (14).

lrytz replied:
aha! thanks!

lrytz reviewed (approved):
Subtle stuff, but LGTM without going through all the details.

adriaanm commented:
I'll push the spec update after lunch. I'd like to get this into M5, but performance fixes will have to wait until RC1.

adriaanm commented:
(The push is just squashing / updating commit comment.)

adriaanm committed:
Spec update: overloading resolution tweaks

Propagate more type information to improve eta-expansion
and function parameter type inference for polymorphic,
higher-order overloaded methods.

adriaanm commented:
I think this is good to go, assuming we are ok with working on regaining the 2% performance drop by RC1.

retronym commented:
Let me take a look at some [profiles](https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/1716/console) to see how likely it is we can reclaim the performance.

retronym commented:
Some initial ideas:

Implement `OverloadedArgFunProto.isHigherKinded`, that's could just be constant false, right?

I see some stack traces like:

```
   scala.collection.immutable.List scala.reflect.internal.Definitions$DefinitionsClass.functionOrPfOrSamArgTypes(scala.reflect.internal.Types$Type) line: 751	1
   boolean scala.reflect.internal.Types$OverloadedArgFunProto.sameHOArgTypes$1(scala.reflect.internal.Types$Type, scala.reflect.internal.Types$Type) line: 1294	1
   scala.Tuple3 scala.reflect.internal.Types$OverloadedArgFunProto.$anonfun$functionArgsProto$2(scala.reflect.internal.Types$OverloadedArgFunProto, scala.Tuple3, scala.Tuple3, scala.reflect.internal.Types$Type) line: 1303	1
   java.lang.Object scala.reflect.internal.Types$OverloadedArgFunProto$$Lambda$394.26866603.apply(java.lang.Object, java.lang.Object) line: -1	1
   java.lang.Object scala.collection.LinearSeqOps.foldLeft(java.lang.Object, scala.Function2) line: 126	1
   java.lang.Object scala.collection.LinearSeqOps.foldLeft$(scala.collection.LinearSeqOps, java.lang.Object, scala.Function2) line: 122	1
   scala.reflect.internal.Types$Type scala.reflect.internal.Types$OverloadedArgFunProto.functionArgsProto$lzycompute() line: 1301	1
   scala.reflect.internal.Types$Type scala.reflect.internal.Types$OverloadedArgFunProto.functionArgsProto() line: 1289	1
   scala.reflect.internal.Types$Type scala.reflect.internal.Types$OverloadedArgFunProto.underlying() line: 1211	1
   boolean scala.reflect.internal.Types$SimpleTypeProxy.isHigherKinded() line: 138	1
   boolean scala.reflect.internal.Types$SimpleTypeProxy.isHigherKinded$(scala.reflect.internal.Types$SimpleTypeProxy) line: 138	1
   boolean scala.reflect.internal.Types$OverloadedArgFunProto.isHigherKinded() line: 1207	1
   boolean scala.reflect.internal.tpe.TypeComparers.isSubType2(scala.reflect.internal.Types$Type, scala.reflect.internal.Types$Type, int) line: 424	1
```

Debug to that point and then figure out what's going on. Is that a cyclic error that's getting caught higher up?

Let's make `samOf` short circuit the `FindMember` walk:

```scala
        val deferredMembers = (
          tp.membersBasedOnFlags(excludedFlags = BridgeAndPrivateFlags, requiredFlags = METHOD).toList.filter(
            mem => mem.isDeferred && !isUniversalMember(mem)
          ) // TODO: test
        )

        // if there is only one, it's monomorphic and has a single argument list
        if (deferredMembers.lengthCompare(1) == 0 &&
            deferredMembers.head.typeParams.isEmpty &&
            deferredMembers.head.info.paramSectionCount == 1)
          deferredMembers.head
```

retronym reviewed (approved):
I'm happy to work on the small performance regression performance together after M5. The change itself LGTM.
